FSAE.X/FSAE_spi.c

```diff
@@ -0,0 +1,336 @@
+  SPI5CONbits.MODE32 = size & 0x20;  // 32/16-Bit Communication Select bits (8-bit)
+  SPI5CONbits.MODE16 = size & 0x10;  // 32/16-Bit Communication Select bits (16-bit)
+  SPI5CONbits.DISSDI = 0; // receiver to ignore the unused bit slots
+  SPI5CONbits.DISSDO = 0; // transmit unused bit slots with logic level 0
+  SPI5CONbits.MSTEN = 1;   // Master Mode Enable bit (Master mode)
+  SPI5CONbits.CKE = 1;     // SPI Clock Edge Select (Serial output data changes on transition from active clock state to idle clock state)
+  SPI5CONbits.SMP = 0;     // SPI Data Input Sample Phase (Input data sampled at middle of output time)
+  SPI5CONbits.CKP = 1;     // Clock Polarity Select (Idle state for clock is a high level)
+
+  // Enable SPI5 module
+  SPI5CONbits.ON = 1;
+
+  lock_config();
+}
+
+void init_spi6(int mhz, int size) {
+  unlock_config();
+
+  // Initialize SDI6/SDO6 PPS pins
+  CFGCONbits.IOLOCK = 0;
+  TRISFbits.TRISF2 = INPUT;
+  SDI6Rbits.SDI6R = 0b1011; // RPF4
+  TRISFbits.TRISF8 = OUTPUT;
+  RPF8Rbits.RPF8R = 0b1010; // SDO6
+  CFGCONbits.IOLOCK = 1;
+
+  // Initialize SCK6
+  TRISDbits.TRISD15 = OUTPUT;
+
+  // Disable interrupts
+  IEC5bits.SPI6IE = 0;
+  IEC5bits.SPI6RXIE = 0;
+  IEC5bits.SPI6TXIE = 0;
+
+  // Disable SPI6 module
+  SPI6CONbits.ON = 0;
+
+  // Clear receive buffer
+  uint32_t readVal = SPI6BUF;
+
+  // Use standard buffer mode
+  SPI6CONbits.ENHBUF = 0;
+
+  /**
+   * F_SCK = F_PBCLK2 / (2 * (SPI1BRG + 1))
+   * SPI1BRG = (F_PBCLK2/(2*F_SCK))-1
+   */
+
+  // Set the baud rate (see above equation, ceil will round to lower freq)
+  SPI6BRG = ceil((PBCLK2/(2.0*mhz))-1);
+
+  SPI6STATbits.SPIROV = 0;
+
+  SPI6CONbits.MCLKSEL = 0; // Master Clock Enable bit (PBCLK2 is used by the Baud Rate Generator)
+  SPI6CONbits.SIDL = 0;    // Stop in Idle Mode bit (Continue operation in Idle mode)
+  SPI6CONbits.MODE32 = size & 0x20;  // 32/16-Bit Communication Select bits (8-bit)
+  SPI6CONbits.MODE16 = size & 0x10;  // 32/16-Bit Communication Select bits (16-bit)
+  SPI6CONbits.DISSDI = 0; // receiver to ignore the unused bit slots
+  SPI6CONbits.DISSDO = 0; // transmit unused bit slots with logic level 0
+  SPI6CONbits.MSTEN = 1;   // Master Mode Enable bit (Master mode)
+  SPI6CONbits.CKE = 1;     // SPI Clock Edge Select (Serial output data changes on transition from active clock state to idle clock state)
+  SPI6CONbits.SMP = 0;     // SPI Data Input Sample Phase (Input data sampled at middle of output time)
+  SPI6CONbits.CKP = 1;     // Clock Polarity Select (Idle state for clock is a high level)
+
+  // Enable SPI6 module
+  SPI6CONbits.ON = 1;
+
+  lock_config();
+}
+
+uint32_t send_spi1(uint32_t value, uint32_t *cs_lat, int cs_num){
+  uint32_t resp = 0;
+
+  *cs_lat &= ~(1 << cs_num); // Set CS Low
+  SPI1BUF = value;
+  while (!SPI1STATbits.SPIRBF);
+  resp = SPI1BUF;
+  *cs_lat |= 1 << cs_num;
+
+  return resp; 
+}
+
+uint32_t send_spi2(uint32_t value, uint32_t *cs_lat, int cs_num){
```

Comment:
Damn, yeah that sucks. What about this idea:

The `send_spix` functions could become `_send_spix`, denoting that they are "private" functions meant only for use within the library itself. And then you could have a `send_spi(uint8_t module_num, ...)` function that basically just calls one of those specific functions based on the `module_num` param.

The reason I like this is that if you have a file that uses a bunch of `send_spix` calls, and you want to go in later and change which SPI bus it uses, you would just have to change one define e.g. `#define CHIP_SPI_BUS 4` instead of changing every single function call. I think it's cleaner in general. But that's just a suggestion, do what you think is best